docs/rfcs/004-decoration-ordering.md

```diff
@@ -0,0 +1,42 @@
+# Decoration ordering
+
+## Status
+
+Proposed
+
+## Summary
+
+Order block decoration items in the DOM in a deterministic and controllable way.
+
+## Motivation
+
+When multiple block decorations are created at the same screen line, they are inserted into the DOM in an order determined by the sequence of their creation; from oldest to newest when `position` is set to `"before"`, from newest to oldest when `position` is set to `"after"`. While this is deterministic, it is limited: it isn't possible to insert decorations within a sequence of existing ones, and it's difficult to control the order of decorations when creating and destroying and moving markers around an editor.
+
+We hit the need for this in [atom/github#1913](https://github.com/atom/github/pull/1913) when we have a block decoration for multiple consecutive collapsed file patches.
+
+## Explanation
+
+[TextEditor::decorateMarker()](https://atom.io/docs/api/v1.34.0/TextEditor#instance-decorateMarker) accepts an additional `order` parameter in its `decorationParams` argument when `type` is "block". When multiple block or overlay decorations occur at the same screen line, they are ordered within the DOM in increasing "order" value. Block decorations with no `order` property are rendered after those with one, in the order in which they were created, oldest to newest.
+
+## Drawbacks
+
+This is a breaking change for co-located block decorations created with an "after" position - they'll now appear in the reverse order.
+
+When multiple packages create block decorations at the same screen line, they'll need to coordinate their `order` values to have expected behavior. There may not even be a clear, universal answer about how block decorations from distinct packages _should_ be ordered.
```

Comment:
What happens if packages _don't_ coordinate and set things to the same `order`? Is it treated like a priority then, and the ordering within the same `order` falls back to the current first/last added?

Reply:
Oh, yeah, I should clarify that. Yes, I was thinking of falling back to creation order for `order` collisions. (Note that I'm proposing changing this to _always_ be oldest-to-newest as well - I suspect the current behavior is there mostly by accident. If this would break any of your stuff, let me know, that's why this is a request for comments :wink:)